README.md
# im

[![Build Status](https://travis-ci.org/publicclass/im.svg?branch=master)](https://travis-ci.org/publicclass/im)

The Stream-based [ImageMagick](http://www.imagemagick.org/) wrapper
for [Node.JS](http://nodejs.org/).

## Install

```bash
$ npm install im
```

## Usage

```javascript
// resizer.js
var im = require('im');
im(process.stdin)
  .resize(process.argv[2])
  .convert(process.stdout)
```

```bash
$ curl http://placekitten.com/500/500 | node resizer.js 200x200 > 200.png
```

## Test

```
$ git clone https://github.com/publicclass/im.git
$ cd im
$ npm i
$ npm test
```

Then open the `test/output` directory and make sure the results are correct.

__Note:__ To test uploading to S3 the env variables: `S3_KEY`, `S3_SECRET`
and `S3_BUCKET` must be set.

### Licence

(The MIT License)

Copyright (c) 2012 Robert Sköld &lt;robert@publicclass.se&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.4.0 / 2015-05-05
==================

  * Added original tests for ident
  * Fixed ident to parse properly according to fixtures (breaking change)

0.3.1 / 2015-05-04
==================

  * Added travis tests
  * Parse the error message from ImageMagick#convert()
  * Added error code for failed input streams
  * Added identify to the exports.commands
  * Fixed the paths and URLs in the http tests


0.3.0 / 2015-05-04
==================

  * Added ImageMagick#interlace(). To be able to make progressive jpegs/
  * Added ImageMagick#identify()


0.2.3 / 2012-06-19
==================

  * Fixed `ImageMagick#coalesce()`, boolean commands don't need any arguments.


0.2.2 / 2012-06-14
==================

  * Added `ImageMagick#layers()`.
  * Added `ImageMagick#coalesce()`.


0.2.1 / 2012-05-29
==================

  * Added `ImageMagick#define()`.
  * Asserts that the `ImageMagick#quality()`-argument is a number.

0.2.0 / 2012-03-19
==================

  * Added `ImageMagick#autoOrient()`.
  * Added `ImageMagick#extent()`.
  * Added `ImageMagick#filter()`.
  * Added `ImageMagick#gravity()`.
  * Added `ImageMagick#liquidRescale()`.
  * Added `ImageMagick#quality()`.
  * Added `ImageMagick#rotate()`.
  * Added `ImageMagick#sample()`.
  * Added `ImageMagick#scale()`.
  * Added `ImageMagick#sharpen()`.
  * Added `ImageMagick#strip()`.
  * Added `ImageMagick#thumbnail()`.
  * Added `ImageMagick#unsharp()`.

0.1.1 / 2012-02-29
==================

  * Added `ImageMagick#format()`.
  * Loosened node engine dependency.
  * Using [slaskis knox fork](https://github.com/slaskis/knox) instead to get
    the streaming to work more reliably.

0.1.0 / 2012-02-29
==================

  * Initial release


test/test-im.js
/*eslint-env mocha */
'use strict';

var im = require('../');
var fs = require('fs');
var path = require('path');
var assert = require('assert');
var request = require('superagent');
var s3;


// only run S3 tests when ENV keys are set
if( process.env.S3_KEY && process.env.S3_SECRET && process.env.S3_BUCKET ){
  var S3FS = require('s3fs');
  s3 = new S3FS(process.env.S3_BUCKET, {
    accessKeyId: process.env.S3_KEY,
    secretAccessKey: process.env.S3_SECRET
  });
}


describe('ImageMagick', function() {

  var dir = path.resolve(__dirname, 'output');

  before(function(next) {
    fs.mkdir(dir, function() { next(); });
  });

  describe('http', function() {
    var url = 'http://ia600604.us.archive.org/21/items/5574288223_f1de5c4692_o/5574288223_f1de5c4692_o_thumb.jpg';

    it('should crop', function(done) {
      var input = request.get(url)
        , output = fs.createWriteStream(path.resolve(dir, 'http_crop_40x40.png'));
      im(input)
        .crop('40x40+90+90')
        .convert(output)
        .on('end', done);
    });

    it('should resize (w. pipe)', function(done) {
      var input = request.get(url)
        , output = fs.createWriteStream(path.resolve(dir, 'http_resize_200x200.png'));
      im(input)
        .resize('200x200')
        .convert()
        .pipe(output)
        .on('close', done);
    });

    it('should crop and resize', function(done) {
      var input = request.get(url)
        , output = fs.createWriteStream(path.resolve(dir, 'http_crop_40x40_resize_200x200.png'));
      im(input)
        .crop('40x40+90+90')
        .resize('200x200')
        .convert(output)
        .on('end', done);
    });

    it('should identify', function(done) {
      var input = request.get(url);
      im(input)
        .identify(function(err, ident) {
          if (err) {
            return done(err);
          }
          assert(typeof ident === 'object');
          assert(typeof ident.gamma === 'number');
          assert(ident.verbose === true);
          done(err);
        });
    });

    it('should fail on non-image streams', function(done) {
      var input = request.get('http://httpbin.org/get')
        , output = fs.createWriteStream(path.resolve(dir, 'http_json_fail.png'));
      im(input)
        .crop('40x40+90+90')
        .resize('200x200')
        .convert(output)
        .on('error', function(err) {
          assert(err.message === 'no images defined');
          done();
        });
    })

    if( !s3 ) {
      return;
    }

    it('should send to s3', function(done) {
      var input = request.get(url)
        , output = s3.createWriteStream('/http_crop_40x40_resize_200x200.png');
      im(input)
        .crop('40x40+90+90')
        .resize('200x200')
        .convert(output)
        .on('end', done);
    });

  });

  describe('fs', function() {
    var file = path.resolve(__dirname, '400.jpg');

    it('should crop', function(done) {
      var input = fs.createReadStream(file)
        , output = fs.createWriteStream(path.resolve(dir, 'file_crop_40x40.jpg'));
      im(input)
        .crop('40x40+90+90')
        .convert(output)
        .on('end', done);
    });

    it('should resize', function(done) {
      var input = fs.createReadStream(file)
        , output = fs.createWriteStream(path.resolve(dir, 'file_resize_200x200.jpg'));
      im(input)
        .resize('200x200')
        .convert(output)
        .on('end', done);
    });

    it('should crop and resize', function(done) {
      var input = fs.createReadStream(file)
        , output = fs.createWriteStream(path.resolve(dir, 'file_resize_200x200_crop_40x40.jpg'));
      im(input)
        .crop('40x40+90+90')
        .resize('200x200')
        .convert(output)
        .on('end', done);
    });

    it('should make a progressive JPEG', function(done) {
      var input = fs.createReadStream(file)
        , output = fs.createWriteStream(path.resolve(dir, 'file_progressive.jpg'));
      im(input)
        .interlace('plane')
        .convert(output)
        .on('end', done);
    });

    it('should identify', function(done) {
      var input = fs.createReadStream(file);
      im(input)
        .identify(function(err, ident) {
          assert(typeof ident === 'object');
          assert(typeof ident.gamma === 'number');
          assert(typeof ident.quality === 'number');
          assert(ident.verbose === true);
          done(err);
        });
    });

    if( !s3 ) {
      return;
    }

    it('should send to s3 (w. pipe)', function(done) {
      var input = fs.createReadStream(file)
        , output = s3.createWriteStream('/file_crop_40x40_resize_200x200.png');
      im(input)
        .crop('40x40+90+90')
        .resize('200x200')
        .convert()
        .pipe(output)
          .on('finish', function() { done(); });
    });

  });

  if (!s3) {
    return;
  }

  describe('s3', function() {

    var file = '/file_crop_40x40_resize_200x200.png';

    before(function(done) {
      s3.exists(file).then(function(exists) {
        if (!exists) {
          return new Promise(function(resolve, reject) {
            var local = path.resolve(dir, 'file_resize_200x200_crop_40x40.jpg');
            fs.readFile(local, function(err, data) {
              if (err) { return reject(err); }
              resolve(data);
            });
          }).then(function(data) {
            return s3.writeFile(file, data);
          });
        } else {
          return Promise.resolve();
        }
      }).then(done, done);
    });

    it('should crop', function(done) {
      var input = s3.createReadStream(file);
      var output = fs.createWriteStream(path.resolve(dir, 's3_crop_40x40.jpg'));
      im(input)
        .crop('40x40+90+90')
        .convert(output)
        .on('end', done);
    });

    it('should resize', function(done) {
      var input = s3.createReadStream(file);
      var output = fs.createWriteStream(path.resolve(dir, 's3_resize_200x200.jpg'));
      im(input)
        .resize('200x200')
        .convert(output)
        .on('end', done);
    });

    it('should crop and resize', function(done) {
      var input = s3.createReadStream(file);
      var output = fs.createWriteStream(path.resolve(dir, 's3_resize_200x200_crop_40x40.jpg'));
      im(input)
        .crop('40x40+90+90')
        .resize('200x200')
        .convert(output)
        .on('end', done);
    });

    it('should send to s3 (w. pipe)', function(done) {
      var input = s3.createReadStream(file);
      var output = s3.createWriteStream('/s3_crop_40x40_resize_200x200.png');
      im(input)
        .crop('40x40+90+90')
        .resize('200x200')
        .convert()
        .pipe(output)
          .on('finish', function(){ done(); });
    });

    it('should identify', function(done) {
      var input = s3.createReadStream(file);
      im(input)
        .identify(function(err, ident) {
          assert(typeof ident === 'object');
          assert(typeof ident.gamma === 'number');
          assert(typeof ident.quality === 'number');
          assert(ident.verbose === true);
          done(err);
        });
    });


  });

});


test/test-ident.js
/*eslint-env mocha*/
'use strict';

var ident = require('../lib/ident');
var path = require('path');
var fs = require('fs');

var jpg = fs.readFileSync(path.resolve(__dirname, 'fixtures/identify-jpg.txt'), 'utf8');
var pdf = fs.readFileSync(path.resolve(__dirname, 'fixtures/identify-pdf.txt'), 'utf8');
var gif = fs.readFileSync(path.resolve(__dirname, 'fixtures/identify-gif.txt'), 'utf8');
var err = fs.readFileSync(path.resolve(__dirname, 'fixtures/identify-err.txt'), 'utf8');
var con = fs.readFileSync(path.resolve(__dirname, 'fixtures/convert-err.txt'), 'utf8');

describe('ident', function() {

  it('should parse a jpg', function() {
    var obj = ident.parse(jpg);
    obj.should.be.a('object');
    obj.should.have.property('image', '/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-AvR72l6Z');
    obj.should.have.property('baseFilename', '-');
    obj.should.have.property('format').include('JPEG');
    obj.should.have.property('geometry');
    obj.should.have.property('type');
    obj.should.have.property('resolution');
    obj.should.have.property('properties');
    obj.properties.should.have.property('dateCreate');
    obj.properties.should.have.property('dateModify');
  });

  it('should parse a pdf', function() {
    var obj = ident.parse(pdf);
    obj.should.be.a('object');
    obj.should.have.property('image', '/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-ggBPXvKO');
    obj.should.have.property('baseFilename', '-');
    obj.should.have.property('format').include('PDF');
    obj.should.have.property('geometry');
    obj.should.have.property('type');
    obj.should.have.property('resolution');
    obj.should.have.property('colors');
    obj.should.have.property('properties');
    obj.properties.should.have.property('dateCreate');
    obj.properties.should.have.property('dateModify');
  });

  it('should parse an animated gif', function() {
    var arr = ident.parse(gif);
    arr.should.be.instanceof(Array);
    arr.should.have.length(5);
    arr.forEach(function(obj) {
      obj.should.be.a('object');
      obj.should.have.property('image', 'original.gif');
      obj.should.have.property('format').include('GIF (CompuServe graphics interchange format)');
      obj.should.have.property('geometry');
      obj.should.have.property('type');
      obj.should.have.property('resolution');
      obj.should.have.property('colors');
      obj.should.have.property('properties');
      obj.properties.should.have.property('dateCreate');
      obj.properties.should.have.property('dateModify');
    });
  });

  it('should parse a "Corrupt data"-error string', function() {
    var e = ident.error(err);
    e.should.have.property('status', 422);
    e.should.have.property('code', 'corrupt');
    e.should.have.property('message', 'Premature end of JPEG file');
    e.should.have.property('description', 'Corrupt JPEG data: premature end of data segment');
    e.should.have.property('filename', '/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-fECuVcfB');
  });

  it('should parse a "invalid argument"-error string', function() {
    var e = ident.error(con);
    e.should.have.property('status', 422);
    e.should.have.property('code', 'invalid_argument');
    e.should.have.property('message', 'invalid argument for option `-resize\': 800+0+0');
    // e.should.have.property('description', 'no decode delegate for this image format') // TODO this should probably be more descriptive...
    e.should.have.property('filename', '/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-mI6bV7wa');
  });

});


lib/im.js
'use strict';

/**
 * Module dependencies.
 */
var Stream = require('stream');
var spawn = require('child_process').spawn;
var ident = require('./ident');


/**
 * Export.
 */
module.exports = exports = ImageMagick;


/**
 * Define the `convert` and `identify` commands path. Defaults to `convert`
 * and `identify` (i.e. relying on the PATH env variable).
 */
exports.commands = {
  convert: 'convert',
  identify: 'identify'
};


/**
 * Stream based imagemagick wrapper.
 *
 * Takes a `Readable Stream` as and puts it through a `convert` process.
 *
 * @param {ReadableStream} input
 */
function ImageMagick(input){
  if( !(this instanceof ImageMagick) ) {
    return new ImageMagick(input);
  }

  if( !input || !input.on ) {
    throw new Error('input must be a readable stream');
  }

  if (input.pause) {
    input.pause();
  }
  this.input = input;
  this.output = '-';
  this.args = [];
}


ImageMagick.prototype = {

  /**
   * Inherits from Stream
   */
  __proto__: Stream.prototype,

  /**
   * Auto Orient.
   *
   * Adds a `-auto-orient` argument. `arg` is a _boolean_.
   *
   * @param {Boolean} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#auto-orient
   * @returns {ImageMagick}
   */
  autoOrient: function(arg){
    if( arg || arg === undefined ) {
      this.args.push('-auto-orient');
    }
    return this;
  },

  /**
   * Coalesce.
   *
   * Adds a `-coalesce` argument. `arg` is a _boolean_.
   *
   * @param {Boolean} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#coalesce
   * @returns {ImageMagick}
   */
  coalesce: function(arg){
    if( arg || arg === undefined ) {
      this.args.push('-coalesce');
    }
    return this;
  },

  /**
   * Crop.
   *
   * Adds a `-crop` argument. `arg` must be an _Image Geometry_ string.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-processing.php#geometry
   * @see http://www.imagemagick.org/script/command-line-options.php#crop
   * @returns {ImageMagick}
   */
  crop: function(arg){
    this.args.push('-crop', arg);
    return this;
  },

  /**
   * Define.
   *
   * Adds a `-define` argument. `arg` must be a _setting_. See docs for more info.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#define
   * @returns {ImageMagick}
   */
  define: function(arg){
    this.args.push('-define', arg);
    return this;
  },

  /**
   * Extent.
   *
   * Adds an `-extent` argument. `arg` must be an _Image Geometry_ string.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-processing.php#geometry
   * @see http://www.imagemagick.org/script/command-line-options.php#extent
   * @returns {ImageMagick}
   */
  extent: function(arg){
    this.args.push('-extent', arg);
    return this;
  },

  /**
   * Filter.
   *
   * Adds a `-filter` argument. `arg` must be a _type_. See docs for more info.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#filter
   * @returns {ImageMagick}
   */
  filter: function(arg){
    this.args.push('-filter', arg);
    return this;
  },

  /**
   * Gravity.
   *
   * Adds a `-gravity` argument. `arg` must be an _gravity type_ string, which
   * means any of `NorthWest`, `North`, `NorthEast`, `West`, `Center`, `East`,
   * `SouthWest`, `South` or `SouthEast`. Defaults to `NorthWest`.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#gravity
   * @returns {ImageMagick}
   */
  gravity: function(arg){
    this.args.push('-gravity', arg);
    return this;
  },

  /**
   * Interlace.
   *
   * Adds a `-interlace` argument. `arg` must be an _interlace type_ string,
   * which means any of `none`, `line`, `plane`, `partition`, `JPEG`, `GIF` or
   * `PNG`. Defaults to `none`.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#interlace
   * @returns {ImageMagick}
   */
  interlace: function(arg){
    this.args.push('-interlace', arg);
    return this;
  },

  /**
   * Layers.
   *
   * Adds a `-layers` argument. `arg` must be a _string_.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#layers
   * @returns {ImageMagick}
   */
  layers: function(arg){
    this.args.push('-layers', arg);
    return this;
  },

  /**
   * Liquid Rescale.
   *
   * Adds a `-liquid-rescale` argument. `arg` must be an _Image Geometry_ string.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-processing.php#geometry
   * @see http://www.imagemagick.org/script/command-line-options.php#liquid-rescale
   * @returns {ImageMagick}
   */
  liquidRescale: function(arg){
    this.args.push('-liquid-rescale', arg);
    return this;
  },

  /**
   * Quality.
   *
   * Adds a `-quality` argument. `arg` is a number between `1` and `100` for JPEGs
   * (default is based on input quality or `92`) and between `1` and `10` for
   * PNGs (defaults to `7`).
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#quality
   * @returns {ImageMagick}
   */
  quality: function(arg){
    if (!isNaN(arg)) {
      this.args.push('-quality', arg);
    }
    return this;
  },

  /**
   * Resize.
   *
   * Adds a `-resize` argument. `arg` must be an _Image Geometry_ string.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-processing.php#geometry
   * @see http://www.imagemagick.org/script/command-line-options.php#resize
   * @returns {ImageMagick}
   * @returns {ImageMagick}
   */
  resize: function(arg){
    this.args.push('-resize', arg);
    return this;
  },

  /**
   * Rotate.
   *
   * Adds a `-rotate` argument. `arg` is the degrees with an optional `<` or- `>`
   * -suffix which makes it only rotate depending on the width/height ratio.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#rotate
   * @returns {ImageMagick}
   */
  rotate: function(arg){
    this.args.push('-rotate', arg);
    return this;
  },

  /**
   * Sample.
   *
   * Adds a `-sample` argument. `arg` must be an _Image Geometry_ string.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-processing.php#geometry
   * @see http://www.imagemagick.org/script/command-line-options.php#sample
   * @returns {ImageMagick}
   */
  sample: function(arg){
    this.args.push('-sample', arg);
    return this;
  },

  /**
   * Scale.
   *
   * Adds a `-scale` argument. `arg` must be an _Image Geometry_ string.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-processing.php#geometry
   * @see http://www.imagemagick.org/script/command-line-options.php#scale
   * @returns {ImageMagick}
   */
  scale: function(arg){
    this.args.push('-scale', arg);
    return this;
  },

  /**
   * Sharpen.
   *
   * Adds a `-sharpen` argument. `arg` is the radius of the Gaussian operator or
   * radius plus sigma. Ex. `radiusxsigma+bias`.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#sharpen
   * @returns {ImageMagick}
   */
  sharpen: function(arg){
    this.args.push('-sharpen', arg);
    return this;
  },

  /**
   * Strip.
   *
   * Adds a `-strip` argument. `arg` is a _boolean_.
   *
   * @param {Boolean} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#strip
   * @returns {ImageMagick}
   */
  strip: function(arg){
    if( arg || arg === undefined ) {
      this.args.push('-strip');
    }
    return this;
  },

  /**
   * Thumbnail.
   *
   * Adds a `-thumbnail` argument. `arg` must be an _Image Geometry_ string.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-processing.php#geometry
   * @see http://www.imagemagick.org/script/command-line-options.php#thumbnail
   * @returns {ImageMagick}
   */
  thumbnail: function(arg){
    this.args.push('-thumbnail', arg);
    return this;
  },

  /**
   * Unsharp.
   *
   * Adds a `-unsharp` argument. `arg` is the radius of the Gaussian operator or
   * radius plus sigma. Ex. `radiusxsigma{+amount}{+threshold}`.
   *
   * @param {String} arg
   * @see http://www.imagemagick.org/script/command-line-options.php#unsharp
   * @returns {ImageMagick}
   */
  unsharp: function(arg){
    this.args.push('-unsharp', arg);
    return this;
  },

  /**
   * Format.
   *
   * Sets the output format.
   *
   * @param {String} format
   * @see `convert -list format`
   * @see
   */
  format: function(format){
    this.output = format + ':-';
    return this;
  },

  /**
   * Convert.
   *
   * Spawns a `convert` process. Pass `output` or `.pipe()` to stream it to a file
   * or something other that handles streams.
   *
   * @param {WritableStream} output
   * @returns {ImageMagick}
   */
  convert: function(output){
    var stderr = '';
    this.proc = spawn(exports.commands.convert, ['-'].concat(this.args, this.output));
    this.proc.stdout.on('data', this.emit.bind(this, 'data'));
    this.proc.stdout.on('end', this.emit.bind(this, 'end'));
    this.proc.stderr.setEncoding('utf8');
    this.proc.stderr.on('data', function(d){ stderr += d; });
    this.proc.stderr.on('end', function() {
      if (stderr) {
        this.emit('error', ident.error(stderr));
      }
    }.bind(this));
    this.proc.on('error', this.emit.bind(this, 'error'));
    this.input.pipe(this.proc.stdin);
    if (this.input.resume) {
      this.input.resume();
    }
    if (output) {
      this.pipe(output);
    }
    return this;
  },

    /**
     * Identify.
     *
     * Spawns an `identify` process. Pass a callback `fn` which will receive
     * a parsed result as an Object.
     *
     * @param {Function} fn
     * @returns {ImageMagick}
     */
  identify: function(fn) {
    var stdout = '';
    var stderr = '';
    this.proc = spawn(exports.commands.identify, ['-verbose', '-']);
    this.proc.stdout.setEncoding('utf8');
    this.proc.stderr.setEncoding('utf8');
    this.proc.stdout.on('data', function(d){ stdout += d; });
    this.proc.stderr.on('data', function(d){ stderr += d; });
    this.proc.on('exit', function(code){
      if (code !== 0 || stderr) {
        fn(ident.error(stderr));
      } else if (stdout) {
        fn(null, ident.parse(stdout));
      } else {
        fn(null, '');
      }
    });
    this.proc.on('error', this.emit.bind(this, 'error'));
    this.input.pipe(this.proc.stdin);
    if (this.input.resume) {
      this.input.resume();
    }
    return this;
  }

};


lib/ident.js
'use strict';

var RE_IDENT = /^((?: {2})*)([^:]+): *([^\n]*)\n*/
  , RE_ERROR = /(identify|convert): *([^@]+?)(?:`([^']+)')? @ (.+?)$/m
  , RE_FIX_KEY = /^(\S+): /
  , RE_CAMEL = /[ -](\w)/ig;

/**
 * Parse the results from an `identify -verbose` command.
 *
 * Converts the indented output of stdout, passed in as `input` into
 * an array of parsed _frame_-objects.
 *
 * @param {String} input
 * @return {Array}
 */
exports.parse = function parse(input){
  // split on each frame
  var frames = input.split(/^(?=I)/gm);
  var idents = frames.map(function(str) {
    var ident = {};
    var stack = [ident];
    var md;

    // trick Image: to be indented like the rest
    str = '  ' + str;

    while ((md = RE_IDENT.exec(str))) {
      var depth = md[1].split('  ').length - 1;
      var key = md[2];
      var val = md[3];

      // fix dates and other keys that contain a colon
      if (RE_FIX_KEY.test(val)) {
        key += ' ' + val.slice(0, RegExp.$1.length);
        val = val.slice(RegExp.$1.length + 2);
      }

      // camel case the keys
      key = key.toLowerCase().replace(RE_CAMEL, toUpper);

      // parse dates into Date
      if (key.indexOf('date') === 0) {
        val = new Date(val);
      } else if (val.toLowerCase() === 'true') {
        val = true;
      } else if (val.toLowerCase() === 'false') {
        val = false;
      } else if (parseFloat(val).toString() === val) {
        val = parseFloat(val);
      } else if (!val) {
        val = undefined;
      }

      while (depth < stack.length) {
        ident = stack.pop();
      }

      // color map uses it's length as val
      // but by ignoring it the colors will
      // be parsed properly
      if (key === 'colormap') {
        val = undefined;
      }

      if (val !== undefined){
        ident[key] = val;
      } else {
        stack.push(ident);
        ident = ident[key] = {};
      }

      str = str.slice(md[0].length);
    }

    return stack[0];
  });

  if (frames.length === 1) {
    return idents[0];
  } else {
    return idents;
  }
};

/**
 * Parse and error from the `identify` command.
 *
 * Example:
 *
 *    identify: Premature end of JPEG file `/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-fECuVcfB' @ warning/jpeg.c/JPEGWarningHandler/325.
 *    identify: Corrupt JPEG data: premature end of data segment `/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-fECuVcfB' @ warning/jpeg.c/JPEGWarningHandler/325.
 *
 *    or convert error
 *
 *    convert: no decode delegate for this image format `/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-SHSPWL4n' @ error/constitute.c/ReadImage/532.
 *    convert: invalid argument for option `-resize': 800+0+0 @ error/convert.c/ConvertImageCommand/2355.
 *
 *  will become something like:
 *
 *    var err = new Error('Premature end of JPEG file')
 *    err.code = 400; // An HTTP status code
 *    err.filename = '/var/folders/_n/pchsjj1d6jqdkd05yjx2tz0c0000gn/T/magick-fECuVcfB';
 *    err.description = 'Corrupt JPEG data: premature end of data segment';
 *    err.raw = str;
 *
 * @param {String} raw
 * @return {Error}
 */
exports.error = function error(raw){
  var err
    , cmd
    , file
    , messages = []
    , md
    , str = raw;

  while ((md = RE_ERROR.exec(str))) {
    cmd = md[1] || cmd;
    file = md[3] || file;

    // Store the messages to create an error
    // with when we figure out the type
    messages.push(md[2].trim().replace(/\\'/g, '\''));

    str = str.slice(md.index + md[0].length);
  }

  // Figure out the code by looking for keywords
  // TODO Add more cases
  switch(true){
    // Error for a corrupt image file.
    case /Corrupt/.test(raw):
      err = new Error(messages.shift());
      err.description = messages.shift();
      err.status = 422;
      err.code = 'corrupt';
      break;

    // Error when trying to identify an xml in which it does into
    // "raw image"-mode in which it requires a size to be set.
    case /must specify image size/.test(raw):
      err = new Error(messages.shift());
      err.description = messages.shift();
      err.status = 422;
      err.code = 'unsupported_type';
      break;

    // Error when there's no input. Usually cause by a failed input stream.
    case /no decode delegate for this image format/.test(raw):

    // Error when the output is not really set, in which it'll just be
    // a `-`.
    case /missing an image filename/.test(raw):

    // Error when trying to convert with an invalid argument.
    // Usually caused by invalid profiles.
    case /invalid argument/.test(raw):
      err = new Error(messages.pop());
      err.description = messages.shift();
      err.status = 422;
      err.code = 'invalid_argument';
      break;

    default:
      console.warn('Could not figure out error code from identify error:');
      console.warn(raw);

      if (!messages.length) {
        console.warn('RE_ERROR did not match the error string');
        err = new Error(raw);
      } else {
        err = new Error(messages.shift());
      }

      err.status = 500;
  }

  // Store command used
  err.command = cmd;

  // Store filename
  err.filename = (file || '').trim().replace(/\\/g, '');

  // Store the original error string
  err.raw = raw;

  return err;
};

function toUpper(_, w){
  return w.toUpperCase();
}
